### PR TITLE
New wizard spell: Forge Arcane Golem

### PR DESCRIPTION
**Pull request by unid15**
http://ss13.moe/wiki/index.php/User:Unid/Forge_Arcane_Golem - read more about spell interactions and other stuff here

triple fireball
bff
- Adds a new spell: Forge Arcane Golem. It requires robes to cast and has a cooldown of 20 seconds (purchasing a cooldown upgrade lowers it to 1)
- Casting the spell creates an arcane golem. It's a fragile (50 health), slow creature that's unable to attack. However, it knows all of its master's spells (except Forge Arcane Golem and Mind Transfer), and will cast them together with the wizard, at the same targets if possible. Spell upgrades aren't transferred to golems, except for controlled fireball (unintended but fine feature)
- Only 1 golem can be summoned at once. Purchasing amount upgrades can raise it up to 3. Casting the spell when there are already enough golems summoned will instead teleport all of your golems to your current location.
- The golem will follow the wizard around and won't cast spells on its own. If it dies, all of the wizard's spell will go on cooldown.
- Sprites by @SkowronX

TODO LIST
- [x] Get somebody to sprite a cool sprite for them
- [x] Fix issues that prevent golem from working at all
- [x] Test #

### CODE REVIEW

**Comment by ComicIronic** [comment]
Wizard just got stronk. Imagine the synergies with Ring of Fire and Magic Missile.

Please restrict this to non-raging.

**Comment by SkowronX** [comment]
heres your damn sprite I'm going home.
animafragment base

**Comment by SkowronX** [comment]
here's your damn sprite I'm going home

animafragment base

NO. I'M MAD. FUCK OFF GOD DAMN JESUS CHRIST.

**Comment by unid15** [comment]
@ComicIronic spell upgrades aren't transfered to golems, so only stationary rings of fire will be spawned. And all of golems' spells (including magic missile and lightning) can harm the wizard, so using offensive golems is kinda risky. Unless you absolutely insist on removing them from ragin mages, I'd rather leave it be and then remove it later if it turns out to be overpowered

@SkowronX thank you

**Comment by SkowronX** [comment]
Oh and here's the original sprite by @FuryMcFlurry originally made for clockcult WHICH DIDNT HAPPEN qq
animafragment base

**Comment by NigglyWiggly** [comment]
> That animation

Skowron, please

**Comment by SkowronX** [comment]
I wonder if anyone noticed I haven't a single fuck in the past six months or so.

NEVERTHELESS. The show must go on!

**Comment by SkowronX** [comment]
should probably get a replacement sprite for it. peeps are saying it's unfitting and I honestly don't really care about this server much anymore.

**Comment by clusterfack** [comment]
I think the purple one is pretty good

File: code/modules/mob/living/simple_animal/friendly/arcane_golem.dm [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+/mob/living/simple_animal/arcane_golem/Die()
+	..()
+
+	//Punish the master by putting all of his spells on cooldown
+	if(master_spell)
+		var/mob/master = master_spell.holder
+
+		if(istype(master))
+			to_chat(master, "<span class='warning'>As your golem is destroyed, your mana leaks away!</span>")
```
Comment (PJB3005) [inline code/modules/mob/living/simple_animal/friendly/arcane_golem.dm:73]: Mana? What?

**Comment by unid15** [comment]
Mana as in "magic energy/magic powers". Less popular definition of the word but it's valid

mana
ˈmɑːnə/
noun
(in Polynesian, Melanesian, and Maori belief) an impersonal supernatural power which can be transmitted or inherited.

**Comment by PJB3005** [comment]
Yes, but I have never seen mana used in SS13 so it just seems weird to use the term.

**Comment by clusterfack** [comment]
Yeah I don't think we've ever established how space wizards actually work

**Comment by NigglyWiggly** [comment]
Well in baylore, I think, it's explained that they have a 'magic organ' that can produce energy with an electrical impulse or something, and NT actively wants them dead because they can harvest these organs to basically just multiply power for free.

In our deep _lore_ I imagine it's just a thing you're trained in using magic energy or some shit

**Comment by FudgePucker** [comment]
Wait does the firebird spell work with them?

**Comment by unid15** [comment]
Yep, each golem summons a firebird

Shame that summon carp isn't in the spellbook, it'd synergise pretty well with this

**Comment by Shadowmech88** [comment]
> Yep, each golem summons a firebird

Do they get into their firebirds when the wizard gets into his?

**Comment by unid15** [comment]
No, they only mimic spellcasting, nothing else

That would be cool however

File: code/modules/spells/spell_code.dm [hunk 2]
```diff
@@ -108,10 +114,16 @@ var/list/spells = typesof(/spell) //needed for the badmin verb for now
 		return (target in options)
 	return ((target in view_or_range(range, user, selection_type)) && istype(target, /mob/living))
 
-/spell/proc/perform(mob/user = usr, skipcharge = 0) //if recharge is started is important for the trigger spells
+/spell/proc/perform(mob/user, skipcharge = 0, list/target_override) //if recharge is started is important for the trigger spells
+	if(!user)
```
Comment (clusterfack) [inline code/modules/spells/spell_code.dm]: This is already implied by the default argument above though?

File: code/modules/spells/spell_code.dm [hunk 3]
```diff
@@ -108,10 +114,16 @@ var/list/spells = typesof(/spell) //needed for the badmin verb for now
...
+		user = usr
+
 	if(!holder)
 		holder = user //just in case
-	if(spell_flags & WAIT_FOR_CLICK)
+
+	var/list/targets = target_override
```
Comment (clusterfack) [inline code/modules/spells/spell_code.dm]: If you do this, there's a similar system in aoe_turf that should be replaced that does an identical thing

File: vgstation13.dme [hunk 4]
```diff
@@ -2036,7 +2039,7 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_tiny.dm"
```
Comment (clusterfack) [inline vgstation13.dme]: Whoops

File: code/modules/mob/living/simple_animal/friendly/arcane_golem.dm [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	var/mob/master = master_spell.holder
+	if(istype(master))
+		if(master.isDead() || isnull(master.loc))
+			Die()
+			return
+
+		//Alive - follow the master
+		if(!isDead())
+			//Don't wander around if wizard is nearby - follow him instead
+			if(step_to(src, master, 2))
```
Comment (clusterfack) [inline code/modules/mob/living/simple_animal/friendly/arcane_golem.dm]: It seems like you want get_dist here

File: code/modules/mob/spells.dm [hunk 6]
```diff
@@ -1,4 +1,7 @@
 /mob/proc/add_spell(var/spell/spell_to_add, var/spell_base = "wiz_spell_ready", var/master_type = /obj/screen/movable/spell_master)
+	if(ispath(spell_to_add, /spell))
```
Comment (clusterfack) [inline code/modules/mob/spells.dm]: Are you sure this distinguishes whether this is actually a path or not?

Comment (ComicIronic) [reply code/modules/mob/spells.dm:2]: What does the second arg for ispath even do?

Comment (clusterfack) [inline code/modules/mob/spells.dm]: It's a lot like istype, except you can feed it two paths.

File: code/modules/mob/living/simple_animal/friendly/arcane_golem.dm (hunk 5 above)
Comment (unid15) [reply code/modules/mob/living/simple_animal/friendly/arcane_golem.dm:60]: step_to returns 0 if the movement failed (can't find a path, or the path's length is smaller than, in this case, 2). I didn't have any issues with this

also butt screw github for removing ability to make line comments on the phone

File: code/modules/mob/spells.dm (hunk 6 above)
Comment (unid15) [reply code/modules/mob/spells.dm:2]: It'll return false if it's an instance or anything that isn't a path (just like istype will return false if fed a null or a path)

--i'll finish this by the end of the weekend - I've been busy lately

**Comment by clusterfack** [comment]
Since github is a shit about removing line comments now, just let me know if/when/which of the issues I have with it are resolved or fixed and then we'll move this along.

File: vgstation13.dme (hunk 4 above)
Comment (unid15) [reply vgstation13.dme]: (fixed)

File: code/modules/spells/spell_code.dm (hunk 3 above)
Comment (unid15) [reply code/modules/spells/spell_code.dm:121]: (done)

File: code/modules/spells/spell_code.dm (hunk 2 above)
Comment (unid15) [reply code/modules/spells/spell_code.dm]: fixed

**Comment by SkowronX** [comment]
What no one wanted to make something to replace that abomination?

**Comment by Exxion** [comment]
What you forget is that your worst sprites are still _leagues_ better than most people's best sprites

**Comment by SonixApache** [comment]
`code\game\gamemodes\wizard\spellbook.dm:39:error: /spell/aoe_turf/conjure/arcane_golem: undefined type path`
`code\modules\spells\general\area_teleport.dm:9:error: AUTOCAST_NOTARGET: undefined var`
AAAAAAAAAAAAAAAAA

**Comment by SkowronX** [comment]
Sucks to be vg.

**Pull request merged**